addons/mass_mailing/controllers/main.py

```diff
@@ -62,6 +62,11 @@ def mailing(self, mailing_id, email=None, res_id=None, token="", **post):
                     'show_blacklist_button': request.env['ir.config_parameter'].sudo().get_param('mass_mailing.show_blacklist_buttons'),
                 })
             else:
+                contacts = request.env['mail.mass_mailing.contact'].sudo().search([('email', '=', email)])
+                subscription_list_ids = contacts.mapped('subscription_list_ids')
+                opt_in_list_ids = subscription_list_ids.filtered(lambda rel: not rel.opt_out).mapped('list_id')
+                unique_opt_in_list_ids = set([list.id for list in opt_in_list_ids])
+                list_ids = request.env['mail.mass_mailing.list'].sudo().browse(unique_opt_in_list_ids)
```

Comment:
browse(unique_opt_in_list_ids) -> You can directly use : browse(opt_in_list_ids.ids)
So you won't need : unique_opt_in_list_ids = set([list.id for list in opt_in_list_ids])

Reply:
My bad, you indeed need a unique list of ids in case you retreive more that one contact subscribed to the same mailign list.